File: clojure/src/aoc/day1.clj
(ns aoc.day1
  (:require [clojure.string :as string]
            [aoc.util :as util]
            [clojure.math.combinatorics :as combo]))

(def input "./../inputs/day1.txt")

(defn read-input []
  (->>
   input
   (slurp)
   (string/split-lines)
   (map util/str->int)))

(defn run-part-1 []
  (->>
   (read-input)
   (#(combo/combinations % 2))
   (filter #(= 2020 (apply + %)))
   (first)
   (apply *)))

(defn run-part-2 []
  (->>
   (read-input)
   (#(combo/combinations % 3))
   (filter #(= 2020 (apply + %)))
   (first)
   (apply *)))

File: elixir/lib/aoc/day1.ex
defmodule Aoc.Day1 do
  alias Aoc.Util

  @input Path.join([__DIR__, "..", "..", "..", "inputs", "day1.txt"])

  def run_part_1 do
    [a, b] =
      read_input()
      |> Util.combinations(2)
      |> Enum.find(fn [a, b] ->
        a + b == 2020
      end)

    IO.puts(a * b)
  end

  def run_part_2 do
    [a, b, c] =
      read_input()
      |> Util.combinations(3)
      |> Enum.find(fn [a, b, c] ->
        a + b + c == 2020
      end)

    IO.puts(a * b * c)
  end

  defp read_input do
    @input
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_integer/1)
  end
end
